Description: the Linux kernel always pads readdir entries, even on 32bit.
Bug: http://sourceware.org/bugzilla/show_bug.cgi?id=11333
Bug-Ubuntu: https://launchpad.net/bugs/392501
Origin: http://repo.or.cz/w/glibc.git/commitdiff/1a81139728494810f65aaa0d0c538ff8c2783dd5

Index: eglibc-2.10.1/sysdeps/unix/readdir_r.c
===================================================================
--- eglibc-2.10.1.orig/sysdeps/unix/readdir_r.c	2010-05-03 15:00:47.184052538 -0700
+++ eglibc-2.10.1/sysdeps/unix/readdir_r.c	2010-05-03 15:01:18.301553857 -0700
@@ -1,4 +1,4 @@
-/* Copyright (C) 1991,92,93,94,95,96,97,98,99,2000,02
+/* Copyright (C) 1991,92,93,94,95,96,97,98,99,2000,02,10
 	Free Software Foundation, Inc.
    This file is part of the GNU C Library.
 
@@ -113,7 +113,17 @@
   while (dp->d_ino == 0);
 
   if (dp != NULL)
-    *result = memcpy (entry, dp, reclen);
+    {
+#ifdef GETDENTS_64BIT_ALIGNED
+      /* The d_reclen value might include padding which is not part of
+	 the DIRENT_TYPE data structure.  */
+      reclen = MIN (reclen, sizeof (DIRENT_TYPE));
+#endif
+      *result = memcpy (entry, dp, reclen);
+#ifdef GETDENTS_64BIT_ALIGNED
+      entry->d_reclen = reclen;
+#endif
+    }
   else
     *result = NULL;
 
Index: eglibc-2.10.1/sysdeps/unix/sysv/linux/i386/readdir64_r.c
===================================================================
--- eglibc-2.10.1.orig/sysdeps/unix/sysv/linux/i386/readdir64_r.c	2010-05-03 15:00:51.785303120 -0700
+++ eglibc-2.10.1/sysdeps/unix/sysv/linux/i386/readdir64_r.c	2010-05-03 15:01:18.301553857 -0700
@@ -1,4 +1,4 @@
-/* Copyright (C) 2000, 2004 Free Software Foundation, Inc.
+/* Copyright (C) 2000, 2004, 2010 Free Software Foundation, Inc.
    This file is part of the GNU C Library.
 
    The GNU C Library is free software; you can redistribute it and/or
@@ -19,6 +19,7 @@
 #define __READDIR_R __readdir64_r
 #define __GETDENTS __getdents64
 #define DIRENT_TYPE struct dirent64
+#define GETDENTS_64BIT_ALIGNED 1
 
 #include <sysdeps/unix/readdir_r.c>
 
